scripts/system_performance/cycle_webscraper.py

```diff
@@ -0,0 +1,198 @@
+# SPDX - License - Identifier: GPL-3.0-or-later
+# ############################################################################### #
+
+"""
+The purpose of this package is for performing MySQL queries to monitor system state
+performance and health. This file retrieves and cleans cycle data for
+use in MySQL queries.
+"""
+
+# Dependencies
+# Web-scraping dependencies
+import os
+import logging
+import re
+import datetime
+import requests
+import lxml.html as lh
+import pandas as pd
+
+
+class TableWebScraper(object):
+    """
+    Web-scrapes beam cycle data from https://www.isis.stfc.ac.uk/Pages/Beam-Status.aspx
+    for MySQL system checks. This URl can be changed at the bottom of the script
+    """
+
+    def __init__(self, url, local_data=None):
+        self.url = url
+        if not local_data:
+            # By default, a local copy of cycle data is stored in the current working directory
+            self.local_data = "cycle_dates.csv"
+        else:
+            # Allows for a custom location to be set for a local CSV copy of cycle data
+            self.local_data = local_data
+
+    def save_to_csv(self, data):
+        """ Create local copy of dataframe as .csv to read when web page is not accessible"""
+        data.to_csv(r'{}'.format(self.local_data), encoding='utf-8')
+        return data
+
+    @staticmethod
+    def read_csv(local_data):
+        """Reads csv into a pandas data frame """
+        if os.path.isfile("{}".format(local_data)):
+            # Directory exists
+            data = pd.read_csv("{}".format(local_data), index_col=0, encoding='utf8')
+            return data
+        # Directory doesn't exist
+        return logging.error("No file named %s found", local_data)
+
+    # -/////////////////////////////////////// Web scraping //////////////////////////////////-#
+
+    def create_table(self):
+        """Private function to format data scraped ready to be placed into a pandas data frame"""
+
+        def _create_header_cols(elements, column):
+            """
+            :param elements: data in each row formatted as nested list
+            :param column: Empty list to be used as header
+            """
+            # For each row, store each first element (header) and an empty list col
+            for text in elements[0]:
+                name = text.text_content()
+                column.append((name, []))
+
+        def _populate_table(elements, col):
+            """Populate  table with web scraped data"""
+
+            # Data is stored on the second row onwards
+            for j in range(1, len(elements)):
+                # tr_elem is our j'th row
+                tr_elem = elements[j]
+
+                # If row is not of size 5, the //tr data is not from table we expect
+                if len(tr_elem) != 5:
+                    break
+
+                # Iterate through each element of the row
+                index = 0
+                for text in tr_elem.iterchildren():
```

Comment:
you can simplify this with using enumerate: https://stackoverflow.com/questions/522563/accessing-the-index-in-for-loops